### PR TITLE
Update infer and execute API to take prompts from txt file for BS>=1

### PR DESCRIPTION
**Pull request by quic-mamta**
Update infer and execute API to take prompts from txt file for BS>=1
If user has passed BS=1 and multiple prompts and wants to execute all of them, that is also supported.

1. User can use --prompt or --prompt_txt_file_path, any of these options.
2. With --prompt, multiple prompts to be passed using "|" delimiter
3. Atleast one argument, either --prompt or --prompt_txt_file_path is required.
4. If any one argument is given then it will take inputs from given argument.
5. If both options are given then inputs will be taken from txt file and a warning is given for same.
6. For BS>1, if number of prompts provided in txt file or prompt don't match BS, it throws an error saying mismatch.
7. We currently don't support running in a loop for BS>1

Example commands : 

1. for BS=1,
    python -m QEfficient.cloud.infer --model-name gpt2 --batch-size 1 --prompt-len 32 --ctx-len 128 --num-cores 16 -- 
    device-group [0] --prompt "My name is"

2. for BS>1,
    python -m QEfficient.cloud.infer --model-name gpt2 --batch-size 3 --prompt-len 32 --ctx-len 128 --num-cores 16 -- 
    device-group [0] --prompt "My name is | The sun rises from | The flat earth theory is the belief that"

3. for BS >=1,
    python -m QEfficient.cloud.infer --model-name gpt2 --batch-size 3 --prompt-len 32 --ctx-len 128 --num-cores 16 -- 
    device-group [0] --prompts_txt_file_path examples/prompts.txt

### CODE REVIEW

**Comment by anujgupt-github** [comment]
Regarding the description: 
"
for bs>1,
python -m QEfficient.cloud.infer --model-name gpt2 --batch-size 3 --prompt-len 32 --ctx-len 128 --num-cores 16 --device-
group '[0]' --inputs_file_path examples/prompts.txt
"

Can a user still use prompts.txt for BS=1, if there's just a single prompt in the txt? If not, that should be allowed too,

**Comment by quic-mamta** [comment]
> Regarding the description: " for bs>1, python -m QEfficient.cloud.infer --model-name gpt2 --batch-size 3 --prompt-len 32 --ctx-len 128 --num-cores 16 --device- group '[0]' --inputs_file_path examples/prompts.txt "
> 
> Can a user still use prompts.txt for BS=1, if there's just a single prompt in the txt? If not, that should be allowed too,

Yes, prompts.txt works with bs 1 also.

**Comment by anujgupt-github** [comment]
LGTM

File: QEfficient/cloud/execute.py [hunk 1]
```diff
@@ -38,7 +39,29 @@ def main(
...
     tokenizer = AutoTokenizer.from_pretrained(model_hf_path, use_cache=True, padding_side="left")
 
-    cloud_ai_100_exec_kv(tokenizer=tokenizer, qpc=qpc_path, device_id=devices, prompt=prompt)
+    if inputs_file_path is not None:
+        try:
+            prompt = []
+            with open(inputs_file_path, "r") as file:
+                for line in file:
+                    prompt.append(line.strip())
+        except FileNotFoundError:
```
Comment (ochougul) [inline QEfficient/cloud/execute.py:48]: Can we add one more except to handle errors that are not `FileNotFoundErrors`, like let's say currupt file or `keyboardinterrupt` error?
Or remove this try except and let the program fail with `FileNotFound` error message?

File: QEfficient/cloud/execute.py [hunk 2]
```diff
@@ -38,7 +39,29 @@ def main(
...
+            print("inputs file not found.")
+
+    qpc_dir_name = qpc_path.strip("/").split("/")[-2]
+    compilation_batch_size = int(qpc_dir_name.split("BS")[0].split("_")[-1])
```
Comment (ochougul) [inline QEfficient/cloud/execute.py:52]: We need to handle this differently, It is possible that user used compile API and provided a different directory name than our standard way of naming qpc directory.

**Review by ochougul — CHANGES_REQUESTED** [review]
Can you think of doing a function that takes in `input_file_path`, `qpc_path`, `prompt` as input and decides how to run the qpc on device based on these three parameters.

This should remove duplication of code in two files, and you can call the same function in two files.

**Comment by ochougul** [comment]
Also, shouldn't we support the same feature via low level API i.e. `cloud_ai_100_exec_kv`, I think it will be easier if we let this API take in the params `qpc_path`, `device_id`, `tokenizer` which are mandatory parameters to be passed by user,
 `prompt` or `input_file_path` which will be optional params, but user must pass either one of them.
 Then this handles how to execute, this way, nothing changes in the infer, execute files, we just pass an extra parameter in `cloud_ai_100_exec_kv` and this will handle everything else.
 
 And we support this feature then, via low as well as high level API.

**Comment by ochougul** [comment]
1. Add XOR condition for exclusively passing either prompt or input_file_path
2. Figure out BS from specializations.json in case of execute
3. Move the logic to exec_kv call, so it doesn't print stats multiple time in case of BS=1, and len(prompt)>1, also clean code.

File: QEfficient/cloud/execute.py [hunk 3]
```diff
@@ -38,7 +39,29 @@ def main(
...
+    if inputs_file_path is not None:
+        try:
+            prompt = []
+            with open(inputs_file_path, "r") as file:
+                for line in file:
+                    prompt.append(line.strip())
+        except FileNotFoundError:
+            print("inputs file not found.")
+
+    qpc_dir_name = qpc_path.strip("/").split("/")[-2]
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/cloud/execute.py]: Done.

File: QEfficient/cloud/execute.py (hunk 2 above)
Comment (quic-mamta) [reply QEfficient/cloud/execute.py]: Updated

**Comment by quic-mamta** [comment]
> 1. Add `XOR` condition for exclusively passing either prompt or `input_file_path`
> 2. Figure out BS from `specializations.json` in case of execute
> 3. Move the logic to `exec_kv` call, so it doesn't print stats multiple time in case of `BS=1`, and `len(prompt)>1`, also clean code.

Updated.

File: QEfficient/cloud/execute.py [hunk 4]
```diff
@@ -49,9 +73,14 @@ def main(
...
         "--prompt",
-        type=lambda prompt: prompt.split("|"),
-        default="My name is",
-        help="Input prompt, if executing for batch size>1, pass input promprs in single string but seperate with pipe (|) symbol",
+        type=str,
+        help="Input prompt, if executing for batch size>1, use prompts_txt_file_path flag",
+    )
+    parser.add_argument(
+        "--prompts_txt_file_path",
+        "--prompts-txt-file-path-file-path",
```
Comment (ochougul) [inline QEfficient/cloud/execute.py:81]: match with line 80

File: QEfficient/cloud/execute.py [hunk 5]
```diff
@@ -67,4 +96,12 @@ def main(
         "--hf-token", "--hf_token", default=None, type=str, required=False, help="HF token id for private HF models"
     )
     args = parser.parse_args()
-    main(args.model_name, args.prompt, args.qpc_path, args.device_group, args.cache_dir, args.hf_token)
+    main(
```
Comment (ochougul) [inline QEfficient/cloud/execute.py:99]: pass `**args.__dict__`

File: QEfficient/generation/text_generation_inference.py [hunk 6]
```diff
@@ -61,6 +61,75 @@ def write_io_files(
...
+
+def read_prompts_txt_file(prompts_txt_file_path: str):
+    prompt = []
+    with open(prompts_txt_file_path, "r") as file:
+        for line in file:
+            prompt.append(line.strip())
+    return prompt
+
+
+def cloud_ai_100_exec_kv(
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:96]: we wanted to add `prompt_txt_file_path` as input to this Low level API right?

**Review by ochougul — CHANGES_REQUESTED** [review]

File: QEfficient/cloud/execute.py [hunk 7]
```diff
@@ -11,14 +11,20 @@
...
+    read_prompts_txt_file,
+)
 from QEfficient.utils import hf_download
 from QEfficient.utils.constants import Constants
 
 
 def main(
     model_name: str,
     prompt: str,
+    prompts_txt_file_path: str,
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/cloud/execute.py:27]: updated

File: QEfficient/cloud/execute.py (hunk 5 above)
Comment (quic-mamta) [reply QEfficient/cloud/execute.py:99]: updated

File: QEfficient/generation/text_generation_inference.py (hunk 6 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py:96]: updated

File: QEfficient/cloud/execute.py (hunk 4 above)
Comment (quic-mamta) [reply QEfficient/cloud/execute.py:81]: updated

File: QEfficient/cloud/execute.py [hunk 8]
```diff
@@ -30,15 +31,24 @@ def main(
     :param model_name: str. Hugging Face Model Card name, Example: [gpt2]
     :prompt: str. Sample prompt for the model text generation
     :qpc_path: str.  Path to the generated binary after compilation.
-    :devices: List[int]. Device Ids to be used for compilation. if devices > 1. Multiple Card setup is enabled.
+    :device_group: List[int]. Device Ids to be used for compilation. if len(device_group) > 1. Multiple Card setup is enabled.
```
Comment (ochougul) [inline QEfficient/cloud/execute.py:34]: Add description for `prompts_txt_file_path`

Comment (quic-mamta) [reply QEfficient/cloud/execute.py:33]: done

File: QEfficient/cloud/execute.py [hunk 9]
```diff
@@ -18,29 +18,38 @@
...
+
     if hf_token is not None:
         login(hf_token)
+
     # Download tokenizer along with model if it doesn't exist
-    model_hf_path = hf_download(repo_id=model_name, cache_dir=cache_dir, allow_patterns=["*.json"])
-    tokenizer = AutoTokenizer.from_pretrained(
-        model_hf_path, use_cache=True, padding_side="left", trust_remote_code=True
-    )
+    model_hf_path = hf_download(repo_id=model_name, cache_dir=cache_dir, allow_patterns=["*.json", "*.py"])
```
Comment (ochougul) [inline QEfficient/cloud/execute.py:42]: Add `*token*` in `allow_patterns`, current code won't download `tokenizer.model`

Comment (quic-mamta) [reply QEfficient/cloud/execute.py]: done

File: QEfficient/generation/text_generation_inference.py [hunk 10]
```diff
@@ -210,6 +298,7 @@ def cloud_ai_100_exec_kv(
         print("Total (E2E) inference time is=", round(total_time, 2))
         return
     print()
+
     print("===================== Performance Stats =====================")
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py:309]: done

File: README.md [hunk 11]
```diff
@@ -110,15 +110,15 @@ In summary:
...
+| QEfficient.cloud.infer           |   [click here](#1-use-qefficientcloudinfer)         |  <li>model_name : $\color{green} {Mandatory}$</li> <li>num_cores : $\color{green} {Mandatory}$</li> <li>device_group : $\color{green} {Mandatory}$</li><li>batch_size : Optional [Default-1]</li> <li>prompt_len : Optional [Default-32]</li> <li>ctx_len : Optional [Default-128]</li><li>mxfp6 : Optional </li> <li>hf_token : Optional </li><li>cache_dir : Optional ["cache_dir" in current working directory]</li><li>prompt : Optional</li><li>prompts_txt_file_path : Optional</li><li>*One argument, prompt or prompts_txt_file_path must be passed*</li> |
+| QEfficient.cloud.execute  |     [click here](#2-use-of-qefficientcloudexecute)       |   <li>model_name : $\color{green} {Mandatory}$</li> <li>device_group : $\color{green} {Mandatory}$</li><li>qpc_path : $\color{green} {Mandatory}$</li> <li>cache_dir : Optional ["cache_dir" in current working directory]</li><li>hf_token : Optional </li><li>prompt : Optional</li><li>prompts_txt_file_path : Optional</li> <li>*One argument, prompt or prompts_txt_file_path must be passed*</li>             |
```
Comment (anujgupt-github) [inline README.md:114]: This statement "One argument, prompt or prompts_txt_file_path must be selected "should not be a bullet.
Put an asterix on both options and below the table, explain the reason for asterix.
Pls fix in both places

File: QEfficient/generation/text_generation_inference.py [hunk 12]
```diff
@@ -222,3 +264,109 @@ def cloud_ai_100_exec_kv(
...
+    prompts_txt_file_path: Optional[str] = None,
+    device_id: List[int] = [0],
+    input_len: Optional[int] = None,
+    generation_len: Optional[int] = None,
+    enable_debug_logs: bool = False,
+    stream: bool = True,
+    write_io_dir: Optional[str] = None,
+    automation=False,
+):
+    if prompts_txt_file_path is not None:
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:350]: Create a function for line 350 to 358, that takes `prompt`, `prompt_file_path`, `batch_size` as parameters and returns `prompts` as output, errors out if inputs are wrong or `batch_size` is not matching `len_prompts`,
Reuse this for line 84-93 in `infer.py`

File: QEfficient/cloud/infer.py [hunk 13]
```diff
@@ -75,6 +80,17 @@ def main(
...
     onnx_dir_path = os.path.join(model_card_dir, "onnx")
     onnx_model_path = os.path.join(onnx_dir_path, model_name.replace("/", "_") + "_kv_clipped_fp16.onnx")
+    
+    assert prompt is not None or prompts_txt_file_path is not None, "Please pass atleast one argument either using --prompt or --prompts_txt_file_path"
+    
+    if prompts_txt_file_path is not None:
+        if prompt is not None:
+            logger.warning("Found inputs passed using txt file as well as CLI, taking inputs from given txt file")
+        prompt = read_prompts_txt_file(prompts_txt_file_path)
+    else:
```
Comment (ochougul) [inline QEfficient/cloud/infer.py:90]: combine with line 90 `else if ...`

File: QEfficient/generation/text_generation_inference.py [hunk 14]
```diff
@@ -222,3 +264,109 @@ def cloud_ai_100_exec_kv(
         print("E2E:", round(total_perf, 2), "tok/s")
         print("Total (E2E) inference time is=", round(total_time, 2), "s")
     print("=============================================================")
+
+
+def cloud_ai_100_exec_kv_helper_loop(
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:269]: mis-leading name, it doesn't run in loop when BS>1.

**Review by ochougul — COMMENTED** [review]
everything else LGTM.

File: README.md (hunk 11 above)
Comment (quic-mamta) [reply README.md]: done

**Review by ochougul — APPROVED** [review]

**Pull request merged**